packages/hint-compat-api/src/utils/filter-supports.ts

```diff
@@ -0,0 +1,66 @@
+import difference = require('lodash/difference');
+import intersection = require('lodash/intersection');
+import union = require('lodash/union');
+
+import { getSupported } from '@hint/utils/dist/src/compat/support';
+
+import { parseSupports, Declaration, DeclarationGroup } from './parse-supports';
+
+/**
+ * Recursively filter provided browsers based on support for the given
+ * `Declaration` or `DeclarationGroup` and its children.
+ */
+const filterItem = (item: Declaration | DeclarationGroup, browsers: string[]): string[] => {
+    if ('prop' in item) {
+        const supportsProperty = getSupported({ property: item.prop }, browsers);
+        const supportsValue = supportsProperty && getSupported({ property: item.prop, value: item.value }, supportsProperty);
+
+        return supportsValue || [];
+    }
+
+    switch (item.type) {
+        case 'and':
+            return intersection(...item.nodes.map((child) => {
+                return filterItem(child, browsers);
+            }));
+        case 'or':
+            return union(...item.nodes.map((child) => {
+                return filterItem(child, browsers);
+            }));
+        case 'not':
+            return difference(browsers, filterItem(item.nodes[0], browsers));
+        default:
+            throw new Error('Unrecognized group type');
+    }
+};
+
+/**
+ * Given the params from an `@supports` rule and a list of target browsers,
+ * return a new list containing only the browsers which would apply the rules
+ * within the `@supports` block.
+ *
+ * ```js
+ * filterSupports('(display: grid)', ['edge 15', 'edge 16']); // returns ['edge 16']
+ * ```
+ */
+export const filterSupports = (params: string, browsers: string[]): string[] | null => {
+    const hasAtSupports = getSupported({ rule: 'supports'}, browsers);
+
+    if (!hasAtSupports) {
+        return null;
+    }
+
+    const root = parseSupports(params);
+
+    if (!root) {
+        return null;
+    }
+
+    try {
+        const supported = filterItem(root, hasAtSupports);
+
+        return supported.length ? supported : null;
+    } catch (e) {
+        return null;
```

Comment:
Similar to the parser, I should throw and catch a known error type here. Then other errors can be re-thrown to avoid silencing unexpected issues.